benchmarks/test_latency.py

```diff
@@ -0,0 +1,136 @@
+                        args_defaults={'tokenizer_type': 'AquilaTokenizer',
+                                       'no_load_rng': True,
+                                       'no_load_optim': True})
+    args = get_args()
+    timers = get_timers()
+    print(f"args is {args}")
+    if args.num_layers_per_virtual_pipeline_stage is not None:
+        print("Interleaved pipeline schedule is not yet supported for text generation.")
+        exit()
+    print_rank_0("WARNING: Forcing exit_on_missing_checkpoint to True for text "
+                 "generation.")
+    args.exit_on_missing_checkpoint = True
+
+    # Set up model and load checkpoint
+    model = get_model(model_provider, wrap_with_ddp=False)
+    print(f"get model successfully....")
+
+    if args.load is not None:
+        _ = load_checkpoint(model, None, None)
+    print("loading model successfully....")
+
+    assert len(model) == 1, "Above condition should have caught this"
+    model = model[0]
+
+    tokenizer = get_tokenizer()
+
+    prompt = ["我" * (args.prompt_len - 1)] * args.micro_batch_size
+
+    print("warming up....")
+    generate_and_post_process(
+            model,
+            prompts=prompt,
+            tokens_to_generate=args.generate_len,
+            return_output_log_probs=True,
+            top_k_sampling=args.top_k,
+            top_p_sampling=args.top_p,
+            top_p_decay=0.0,
+            top_p_bound=0.0,
+            temperature=args.temperature,
+            add_BOS=False,
+            use_eod_token_for_early_termination=True,
+            stop_on_double_eol=False,
+            stop_on_eol=False,
+            prevent_newline_after_colon=False,
+            random_seed=args.seed,
+    )
+
+    print("Benchmark....")
+    timers("all_iters", log_level=0).start(barrier=True)
+    for i in range(args.num_iters):
+        generate_and_post_process(
+            model,
+            prompts=prompt,
+            tokens_to_generate=args.generate_len,
+            return_output_log_probs=True,
+            top_k_sampling=args.top_k,
+            top_p_sampling=args.top_p,
+            top_p_decay=0.0,
+            top_p_bound=0.0,
+            temperature=args.temperature,
+            add_BOS=False,
+            use_eod_token_for_early_termination=True,
+            stop_on_double_eol=False,
+            stop_on_eol=False,
+            prevent_newline_after_colon=False,
+            random_seed=args.seed,
+        )
+    time = timers("all_iters").elapsed()
+
+    device = next(model.parameters()).device
+    memory_used = torch.cuda.max_memory_allocated(device) / (1024 ** 3)
+    memory_pct = memory_used / (torch.cuda.get_device_properties(device).total_memory / (1024 ** 3)) * 100
+
+    print("="*30)
+    print(f"Batch Size: {args.micro_batch_size}")
+    print(f"Prompt len: {args.prompt_len}")
+    print(f"Generate len: {args.generate_len}")
+    print(f"Avg Latency: {time / args.num_iters:.2f} seconds")
```

Comment:
Please add more comments about the difference between the latency and throughput calculation.

Reply:
Done.